homeassistant/components/amcrest/camera.py

```diff
@@ -112,28 +118,58 @@ def __init__(self, name, device, ffmpeg):
         self._motion_recording_enabled = None
         self._color_bw = None
         self._rtsp_url = None
-        self._snapshot_lock = asyncio.Lock()
+        self._snapshot_task = None
         self._unsub_dispatcher = []
         self._update_succeeded = False
 
-    async def async_camera_image(self):
-        """Return a still image response from the camera."""
+    def _check_snapshot_ok(self):
         available = self.available
         if not available or not self.is_on:
             _LOGGER.warning(
                 "Attempt to take snaphot when %s camera is %s",
                 self.name,
                 "offline" if not available else "off",
             )
+            raise _CannotSnapshot
+
+    async def _async_get_image(self):
+        try:
+            # Send the request to snap a picture and return raw jpg data
+            # Snapshot command needs a much longer read timeout than other commands.
+            return await self.hass.async_add_executor_job(
```

Comment:
It looks like we can use this task as the snapshot task instead of wrapping it inside a coroutine.

Reply:
Then who catches any exceptions? The non-async function being called here is in the pypi package, not HA.

Reply:
We move that below inside `async_camera_image`.

Reply:
Now we've come full circle. As I explained earlier, the task that calls `async_camera_image` can get canceled, so...

Reply:
Ok.